### ISSUE TITLE
isal.igzip_lib.IgzipDecompressor does not expose "unconsumed_tail"

### ISSUE THREAD

**Issue by alexander-beedie** [pycompression/python-isal#116]
`IgzipDecompressor` doesn't need `unconsumed_tail` for its own use, but for lower-level gzip module compatibility it would be good to expose it, and have it always return `b""`?

**Comment by rhpvorderman** [comment]
It is designed after the BZ2Decompressor, which also does not have an unconsumed tail.

The reason is that this works better for file decompression (zlib.Decompress works better for in-memory decompression). As it is now, igzip.py has a much smaller overhead than gzip.py when decompressing files. The machinery in gzip.py is quite inefficient. I want to upstream my improvements at some point, but have not found the time.

**Comment by alexander-beedie** [comment]
Yup, no problem with the design; solely looking at for compatibility with lower-level code that expects the decompressor to have an `unconsumed_tail` attribute. Since it isn't really used, I suggest adding one that always returns `b""` (which is typically what is being checked for).

Came up when I contributed a [small patch](https://github.com/pgzip/pgzip/pull/21/files) to `pgzip` just now (adding support for `isal.igzip`), and had to make the following tweak:
```python
    ## before (calling into standard gzip)
    if self._decompressor.unconsumed_tail != b"":
...
    ## after (decompressor may be IgzipDecompressor)
    if getattr(self._decompressor, "unconsumed_tail", b"") != b"":
```
It's obviously not a burden, but might be nice to add the attribute to slightly increase drop-in compatibility  ;)

**Comment by rhpvorderman** [comment]
Have you checked `isal_zlib.decompressobj` ? That does have an unconsumed tail and was made with low level compatibility in mind.

**Comment by alexander-beedie** [comment]
> Have you checked `isal_zlib.decompressobj` ? That does have an unconsumed tail and was made with low level compatibility in mind.

Yep, I saw that was present; however, in the patch above, the underlying code appears imply the decompressor from GzipFile, hence my wondering if adding it for compatibility purposes would be reasonable. The special-case I added is minimally intrusive, so I'm fine with it as-is.

Also, it looks like the compression levels for `isal_zlib.compressobj` don't map 1:1 to the compression levels for `zlib.compressobj` either. Would it be reasonable to do something like map `1-3 => 1`, `4-6 => 2`, and `7-9 => 3` (gzip => igzip)? Otherwise this doesn't really work as a drop-in replacement for `zlib.decompressobj` either. 

I'm not sure to what extent drop-in replacement equivalence is a project goal, so feel free to ignore; happy to use it as-is, accounting for the differences in local code where necessary  ;)

**Comment by rhpvorderman** [comment]
> Also, it looks like the compression levels for isal_zlib.compressobj don't map 1:1 to the compression levels for zlib.compressobj either.

No, if you look at the compression levels, levels 1,2 and 3 deliver very similar results. They compress more like level 4 of the zlib library. Level 0 is completely different it compresses very badly. Still better than zlib 0, which does not compress at all (which makes more sense for 0 in my opinion). 

So it is not really possible to map them to the zlib compression levels. There are no equivalents for the higher levels. That is okay. This library is all about speed, not compression ratio. 

As regards to pgzip, I think you will be better of using [xopen](https://github.com/pycompression/xopen). That library wraps pigz for parallel compression, but falls back on python-isal for the lower compression levels, as well as for decompression. Xopen seems a lot more mature than pgzip. Disclaimer: I have contributed extensively to xopen. 

Also there are a lot of not so untuitive caveats for Python performance. Most of them have been addressed in xopen. I wouldn't be surprised if xopen absolutely crushes pgzip on some performance metrics, but your mileage might vary.
